src/test/java/org/wise/portal/dao/user/impl/HibernateUserDaoTest.java

```diff
@@ -17,186 +17,504 @@
+    assertEquals("MrsPuff", user.getUserDetails().getUsername());
+  }
+
+  @Test
+  public void retrieveTeacherUsersJoinedSinceYesterday_WithNewRecentTeachers_ShouldReturnUsers() {
+    List<User> users = userDao.retrieveTeacherUsersJoinedSinceYesterday();
+    assertEquals(2, users.size());
+  }
+
+  @Test
+  public void retrieveTeacherUsersWhoLoggedInSinceYesterday_NoRecentLogins_ShouldReturnNoUsers() {
+    List<User> users = userDao.retrieveTeacherUsersWhoLoggedInSinceYesterday();
+    assertEquals(0, users.size());
+  }
+
+  @Test
+  public void retrieveTeacherUsersWhoLoggedInSinceYesterday_WithRecentLogin_ShouldReturnUsers() {
+    ((TeacherUserDetails) teacher1.getUserDetails()).setLastLoginTime(new Date());
+    List<User> users = userDao.retrieveTeacherUsersWhoLoggedInSinceYesterday();
+    assertEquals(1, users.size());
+  }
+
+  @Test
+  public void retrieveTeacherUsersWhoLoggedInToday_WithRecentLogin_ShouldReturnUsers() {
+    ((TeacherUserDetails) teacher1.getUserDetails()).setLastLoginTime(new Date());
+    ((TeacherUserDetails) teacher2.getUserDetails()).setLastLoginTime(new Date());
+    List<User> users = userDao.retrieveTeacherUsersWhoLoggedInToday();
+    assertEquals(2, users.size());
+  }
+
+  @Test
+  public void retrieveTeacherUsersWhoLoggedInThisWeek_WithRecentLogin_ShouldReturnUsers() {
+    ((TeacherUserDetails) teacher1.getUserDetails()).setLastLoginTime(getDateXDaysFromNow(-8));
+    Calendar c = Calendar.getInstance();
+    c.set(Calendar.DAY_OF_WEEK, 1);
+    ((TeacherUserDetails) teacher2.getUserDetails())
+        .setLastLoginTime(new Date(c.getTimeInMillis() + 1000));
+    List<User> users = userDao.retrieveTeacherUsersWhoLoggedInThisWeek();
+    assertEquals(1, users.size());
+  }
+
+  @Test
+  public void retrieveTeacherUsersWhoLoggedInThisMonth_WithRecentLogin_ShouldReturnUsers() {
+    ((TeacherUserDetails) teacher1.getUserDetails()).setLastLoginTime(getDateXDaysFromNow(-32));
+    Calendar c = Calendar.getInstance();
+    c.set(Calendar.DAY_OF_MONTH, 1);
+    ((TeacherUserDetails) teacher2.getUserDetails())
+        .setLastLoginTime(new Date(c.getTimeInMillis() + 1000));
+    List<User> users = userDao.retrieveTeacherUsersWhoLoggedInThisMonth();
+    assertEquals(1, users.size());
+  }
+
+  @Test
+  public void retrieveTeacherUsersWhoLoggedInThisYear_WithRecentLogin_ShouldReturnUsers() {
+    ((TeacherUserDetails) teacher1.getUserDetails()).setLastLoginTime(getDateXDaysFromNow(-366));
+    Calendar c = Calendar.getInstance();
+    c.set(Calendar.DAY_OF_YEAR, 1);
+    ((TeacherUserDetails) teacher2.getUserDetails())
+        .setLastLoginTime(new Date(c.getTimeInMillis() + 1000));
+    List<User> users = userDao.retrieveTeacherUsersWhoLoggedInThisYear();
+    assertEquals(1, users.size());
+  }
+
+  @Test
+  public void retrieveStudentUsersWhoLoggedInSinceYesterday_WithRecentLogin_ShouldReturnUsers() {
+    ((StudentUserDetails) student1.getUserDetails()).setLastLoginTime(new Date());
```

Comment:
```suggestion
    student1.getUserDetails().setLastLoginTime(new Date());
```
Same as teacherUserDetails, no need to cast.